### PR TITLE
Rename view

### PR DESCRIPTION
**Pull request by gagalago**
allow to rename views managed by Scenic

- [x] rename views
- [x] definition of views follow the renaming
- [x] ease the flow of replacing views (fix https://github.com/scenic-views/scenic/issues/318)

### CODE REVIEW

**Comment by calebthompson** [comment]
Liking this so far!

**Comment by calebthompson** [comment]
A thought for versioning: can we take the old view version as either a required argument for what file to copy or an optional argument and calculate the current version of it is not given?

Should a renamed view maintain the version number of the original name or start fresh?

**Comment by gagalago** [comment]
> A thought for versioning: can we take the old view version as either a required argument for what file to copy or an optional argument and calculate the current version of it is not given?
we cannot copy the file from scenic gem because if we rename 2 views (ex: A and B) in the same migration and that one of them use the other (ex: A use B), then the resulting SQL file need also that the in the SQL of the depending view, the other is renamed. As most of the time that will not be an issue, I think it's better to copy the file beforehand thanks to a rake generator task.

> Should a renamed view maintain the version number of the original name or start fresh?
if we want to allow flow as described in https://github.com/scenic-views/scenic/issues/318 we are not able to maintain the original version number, so we will have to start fresh (or use the next version available for the new name).

File: lib/scenic/statements.rb [hunk 1]
```diff
@@ -141,15 +141,59 @@ def replace_view(name, version: nil, revert_to_version: nil, materialized: false
...
+    #   by substituing in their name the previous view name
+    #   to the next view name. Defaults to false.
+    # @return The database response from executing the rename statement.
+    #
+    # @example Rename a view
+    #   drop_view(:engggggement_reports, :engagement_reports)
+    #
+    def rename_view(
+      from_name, to_name,
+      version: nil, revert_to_version: nil, materialized: false
```
[the remark this answers is not in the document]
Comment (gagalago) [reply lib/scenic/statements.rb:170]: this is used when we rollback

File: README.md [hunk 2]
```diff
@@ -195,6 +195,50 @@ you would need to refresh view B first, then right after refresh view A. If you
...
+$ rails generate scenic:view results --rename search_results
+      create  db/views/results_v01.sql
+      create  db/migrate/[TIMESTAMP]_update_results_to_version_2.rb
+```
+
+Now, `db/views/results_v01.sql` sould be identical to 
+`db/views/search_results_v01.sql`, you will need to edit this file if you rename
+another views in this migration that is in this definition, so just rename it
+too accordingly ;-)
+The migration should look something like:
```
Comment (woidda) [inline README.md:212]:
```suggestion
The migration should look something like this:
```

File: README.md [hunk 3]
```diff
@@ -195,6 +195,50 @@ you would need to refresh view B first, then right after refresh view A. If you
...
+
+```ruby
+class UpdateResultsToVersion2 < ActiveRecord::Migration
+  def change
+    rename_view :search_results, :results, version: 2, revert_to_version: 1
+  end
+end
+```
+
+## Need to change materialized view without downtime?
```
Comment (woidda) [inline README.md:222]:
```suggestion
## Need to change a materialized view without downtime?
```

File: lib/scenic/adapters/postgres.rb [hunk 4]
```diff
@@ -222,10 +265,53 @@ def refresh_materialized_view(name, concurrently: false, cascade: false)
         end
       end
 
+      # Returns a normalize a SQL query
```
Comment (woidda) [inline lib/scenic/adapters/postgres.rb:268]:
```suggestion
      # Returns a normalized SQL query
```

File: lib/scenic/adapters/postgres.rb [hunk 5]
```diff
@@ -222,10 +265,53 @@ def refresh_materialized_view(name, concurrently: false, cascade: false)
...
+      def normalize_sql(sql_definition)
+        temporary_view_name = "temp_view_for_decompilation"
+        view_name = quote_table_name(temporary_view_name)
+        transaction do
+          execute "CREATE TEMPORARY VIEW #{view_name} AS #{sql_definition};"
+          normalize_view_sql(temporary_view_name)
+        end
+      end
+
+      # Returns a normalize the SQL definition of a view
```
Comment (woidda) [inline lib/scenic/adapters/postgres.rb:284]:
```suggestion
      # Returns a normalized SQL definition of a view
```

File: lib/scenic/command_recorder/statement_arguments.rb [hunk 6]
```diff
@@ -3,11 +3,12 @@ module CommandRecorder
...
     class StatementArguments
       def initialize(args)
-        @args = args.freeze
+        @args = args.clone
+        @options = @args.extract_options!
       end
 
       def view
-        @args[0]
+        args[0]
```
Comment (woidda) [inline lib/scenic/command_recorder/statement_arguments.rb:11]: Why did you remove the "@"?

File: lib/scenic/configuration.rb [hunk 7]
```diff
@@ -6,8 +8,15 @@ class Configuration
     # @return Scenic adapter
     attr_accessor :database
 
+    # The full path where is stored views definition sql files.
```
Comment (woidda) [inline lib/scenic/configuration.rb:11]:
```suggestion
    # The full path for files containing SQL definitions.
```

File: lib/scenic/errors.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
+module Scenic
+  # Raised when a view definition in the database is different
+  # from the definition used to create migrations.
```
Comment (woidda) [inline lib/scenic/errors.rb:3]:
```suggestion
  # from the definition used to create the migrations.
```

File: spec/scenic/definition_spec.rb [hunk 9]
```diff
@@ -32,30 +32,23 @@ module Scenic
 
       it "handles schema qualified view names" do
         definition = Definition.new("non_public.searches", 1)
+        expected = Rails.root.join("db/views/non_public_searches_v01.sql")
```
Comment (woidda) [inline spec/scenic/definition_spec.rb:35]:
```suggestion
        expected = Rails.root.join("db", "views", "non_public_searches_v01.sql")
```

File: spec/scenic/statements_spec.rb [hunk 10]
```diff
@@ -192,6 +192,51 @@ module Scenic
...
+        connection.rename_view(
+          :from, :to,
+          version: 1, materialized: { rename_indexes: true }
+        )
+
+        expect(Scenic.database).to have_received(:rename_materialized_view)
+          .with(:from, :to, rename_indexes: true)
+      end
+
+      it "raises an error if not supplied a version" do
```
Comment (woidda) [inline spec/scenic/statements_spec.rb:223]:
```suggestion
      it "raises an error if no version supplied" do
```

File: spec/scenic/statements_spec.rb [hunk 11]
```diff
@@ -192,6 +192,51 @@ module Scenic
...
+        expect { connection.rename_view(:from, :to) }
+          .to raise_error(ArgumentError, /version is required/)
+      end
+
+      it "raises an error if view in database is not similar to definition" do
```
Comment (woidda) [inline spec/scenic/statements_spec.rb:228]:
```suggestion
      it "raises an error if the view in the database is not similar to definition" do
```

**Review by woidda — APPROVED** [review]
LGTM

**Comment by gagalago** [comment]
@calebthompson @derekprior I am open to comments.

I didn't touch to public documented methods but I changed some private to ease the creation of this feature, Is it fine for you?

I focused on renaming views, do you think that I need to go further to ease replacing materialized views (maybe in a following pull request as that will only used primitives that are present in Scenic and in this pull request)?

File: lib/scenic/statements.rb [hunk 12]
```diff
@@ -123,33 +126,83 @@ def update_view(name, version: nil, sql_definition: nil, revert_to_version: nil,
...
+    # @param materialized [Boolean] True if updating a materialized view.
     # @return The database response from executing the create statement.
     #
     # @example
     #   replace_view :engagement_reports, version: 3, revert_to_version: 2
     #
-    def replace_view(name, version: nil, revert_to_version: nil, materialized: false)
+    def replace_view(
+      from_name, to_name = nil,
+      version: nil, revert_to_version: nil, materialized: false
```
[the remark this answers is not in the document]
Comment (gagalago) [reply lib/scenic/statements.rb:143]: used on revert

**Review by mihaic195 — APPROVED** [review]

**Comment by gagalago** [comment]
in case of a used column renaming, we are not able to replace the view. I will have work again on this pull request to improve it.

**Comment by gagalago** [comment]
I used this branch in production and it worked as expected. 

@calebthompson @derekprior  do you have some others remarks?

**Comment by trevor-trou** [comment]
Are there plans to merge this branch?

**Comment by calebhearth** [comment]
After a brief meeting of the Scenic core team over text, no I don’t believe we’ll be merging this. Rather than renaming, we recommend dropping the old view and creating a new one. 

However, we do see the value in a “zero downtime” materialize view update and would entertain a PR that explored on that more specifically.

We do appreciate the work that went into writing and verifying this, so thank you!

**Comment by trevor-trou** [comment]
Credit to @gagalago, I just noticed this branch while searching for a “zero downtime” materialized view migration solution and was curious about its future.

Thanks for the update!
